### PR TITLE
[network diagnostic] updates to consistent with spec

### PR DESCRIPTION
**Pull request by librasungirl**
Type 3 Timeout - For SEDs, indicates the maximum polling time period. A non-SED Thread device MUST omit this TLV in a DIAG_GET.rsp.
Add Type 19 Max Child Timeout - This field reports the value of the longest MLE Timeout TLV a router has registered for all of its active MTD children

### CODE REVIEW

File: src/core/thread/data_poll_manager.cpp [hunk 1]
```diff
@@ -170,6 +170,23 @@ void DataPollManager::SetExternalPollPeriod(uint32_t aPeriod)
...
+{
+    uint32_t pollperiod = 0;
+
+    if (mExternalPollPeriod != 0)
+    {
+        pollperiod = mExternalPollPeriod;
+    }
+    else
+    {
+        pollperiod = Timer::SecToMsec(GetMeshForwarder().GetNetif().GetMle().GetTimeout()) -
```
Comment (jwhui) [inline src/core/thread/data_poll_manager.cpp]: This code is duplicated from `DataPollManager::CalculatePollPeriod()`.  Shall we separate this computation out into a new method like `DataPollManager::GetDefaultPollPeriod()`?

File: src/core/thread/data_poll_manager.hpp [hunk 2]
```diff
@@ -185,6 +185,14 @@ class DataPollManager: public MeshForwarderLocator
      */
     void SendFastPolls(uint8_t aNumFastPolls);
 
+    /**
+     * This methods gets the sleepy polling period
```
Comment (jwhui) [inline src/core/thread/data_poll_manager.hpp]: May be more correct to say:
> This method gets the maximum data polling period in use.

File: src/core/thread/data_poll_manager.hpp [hunk 3]
```diff
@@ -185,6 +185,14 @@ class DataPollManager: public MeshForwarderLocator
...
+     *
+     * @returns the sleepy polling period
```
Comment (jwhui) [inline src/core/thread/data_poll_manager.hpp]: Before:
> the sleepy polling period

After:
> the maximum data polling period in use

File: src/core/thread/mle_router.cpp [hunk 4]
```diff
@@ -4706,6 +4706,36 @@ MleRouter &MleRouter::GetOwner(const Context &aContext)
...
+    VerifyOrExit(mRole == OT_DEVICE_ROLE_ROUTER || mRole == OT_DEVICE_ROLE_LEADER, error = OT_ERROR_INVALID_STATE);
+
+    for (uint8_t i = 0; i < kMaxChildren; i++)
+    {
+        if (mChildren[i].GetState() != Neighbor::kStateValid)
+        {
+            continue;
+        }
+
+        if (mChildren[i].IsFullThreadDevice())
```
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: If the metric is limited to SEDs, then we probably want to use `IsRxOnWhenIdle()` method instead.

File: src/core/thread/mle_router.cpp [hunk 5]
```diff
@@ -4706,6 +4706,36 @@ MleRouter &MleRouter::GetOwner(const Context &aContext)
     return mle;
 }
 
+otError MleRouter::GetMaxChildTimeout(uint32_t &aTimeout)
+{
+    otError error = OT_ERROR_NOT_FOUND;
+
+    VerifyOrExit(mRole == OT_DEVICE_ROLE_ROUTER || mRole == OT_DEVICE_ROLE_LEADER, error = OT_ERROR_INVALID_STATE);
+
+    for (uint8_t i = 0; i < kMaxChildren; i++)
```
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: Replace `kMaxChildren` with `mMaxChildrenAllowed`?

File: src/core/thread/mle_router_ftd.hpp [hunk 6]
```diff
@@ -684,6 +684,18 @@ class MleRouter: public Mle
      */
     otError SetAssignParentPriority(int8_t aParentPriority);
 
+    /**
+     * This method gets the longest MLE Timeout TLV for all active MTD children.
+     *
+     * @param[in]  aTimeout  A reference to where the information is placed.
```
Comment (jwhui) [inline src/core/thread/mle_router_ftd.hpp]: Specify as output: `@param[out]`

File: src/core/thread/mle_router_ftd.hpp [hunk 7]
```diff
@@ -684,6 +684,18 @@ class MleRouter: public Mle
...
+     *
+     * @retval OT_ERROR_NONE           Successfully get the max child timeout
+     * @retval OT_ERROR_INVALID_STATE  Not an active router
+     * @retval OT_ERROR_NOT_FOUND      NO MTD child
+     *
+     */
+    otError GetMaxChildTimeout(uint32_t &aTimeout);
```
Comment (jwhui) [inline src/core/thread/mle_router_ftd.hpp]: Declare as `const`?

File: src/core/thread/data_poll_manager.hpp [hunk 8]
```diff
@@ -185,6 +185,14 @@ class DataPollManager: public MeshForwarderLocator
...
+     *
+     */
+    uint32_t GetKeepAlivePollPeriod(void);
```
Comment (jwhui) [inline src/core/thread/data_poll_manager.hpp]: Declare as `const`?

File: src/core/thread/mle_router.cpp (hunk 4 above)
Comment (librasungirl) [inline src/core/thread/mle_router.cpp]: this metric is the longest MLE Timeout TLV for all active MTD children. not only sed.

Comment (jwhui) [inline src/core/thread/mle_router.cpp]: A rant on the spec and not this PR itself: while I agree the spec defines this metric for all MTDs, I do not believe this metric is very useful.  Having the maximum timeout for SEDs is especially useful for setting the Pending Operational Dataset Delay Timer value.

**Pull request merged**